zjit/src/codegen.rs

```diff
@@ -646,10 +646,20 @@ fn gen_guard_block_param_proxy(jit: &JITState, asm: &mut Assembler, level: u32,
 }
 
 fn gen_guard_not_frozen(jit: &JITState, asm: &mut Assembler, recv: Opnd, state: &FrameState) -> Opnd {
-    let ret = asm_ccall!(asm, rb_obj_frozen_p, recv);
-    asm_comment!(asm, "side-exit if rb_obj_frozen_p returns Qtrue");
-    asm.cmp(ret, Qtrue.into());
-    asm.je(side_exit(jit, state, GuardNotFrozen));
+    let side_exit = side_exit(jit, state, GuardNotFrozen);
+    let recv = asm.load(recv);
+    // Side-exit if recv is false
+    assert_eq!(Qfalse.as_i64(), 0);
+    asm.test(recv, recv);
+    asm.jz(side_exit.clone());
+    // Side-exit if recv is immediate
+    asm.test(recv, (RUBY_IMMEDIATE_MASK as u64).into());
+    asm.jnz(side_exit.clone());
```

Comment:
This non-immediate check seems to be copied everywhere at the moment. It might be nice to have a helper function that does this. In YJIT, it's called `guard_object_is_heap`.

Reply:
Totally agree